pkg/apis/core/validation/validation.go

```diff
@@ -7246,8 +7247,60 @@ func validateNodeInclusionPolicy(fldPath *field.Path, policy *core.NodeInclusion
 	return nil
 }
 
-// validateMatchLabelKeys tests that the elements are a valid label name and are not already included in labelSelector.
-func validateMatchLabelKeys(fldPath *field.Path, matchLabelKeys []string, labelSelector *metav1.LabelSelector) field.ErrorList {
+// validateMatchLabelKeysAndMismatchLabelKeys checks if both matchLabelKeys and mismatchLabelKeys are valid.
+// - validate that all matchLabelKeys and mismatchLabelKeys are valid label names.
+// - validate that the user doens't specify the same key in both matchLabelKeys and labelSelector.
+// - validate that any matchLabelKeys are not duplicated with mismatchLabelKeys.
+func validateMatchLabelKeysAndMismatchLabelKeys(fldPath *field.Path, matchLabelKeys, mismatchLabelKeys []string, labelSelector *metav1.LabelSelector) field.ErrorList {
+	var allErrs field.ErrorList
+	// 1. validate that all matchLabelKeys and mismatchLabelKeys are valid label names.
+	allErrs = append(allErrs, validateLabelKeys(fldPath.Child("matchLabelKeys"), matchLabelKeys, labelSelector)...)
+	allErrs = append(allErrs, validateLabelKeys(fldPath.Child("mismatchLabelKeys"), mismatchLabelKeys, labelSelector)...)
+
+	// 2. validate that the user doens't specify the same key in both matchLabelKeys and labelSelector.
+	// It doesn't make sense to have the labelselector with the key specified in matchLabelKeys
+	// because the matchLabelKeys will be `In` labelSelector which matches with only one value in the key
+	// and we cannot make any further filtering with that key.
+	// On the other hand, we may want to have labelSelector with the key specified in mismatchLabelKeys.
+	// because the mismatchLabelKeys will be `NotIn` labelSelector
+	// and we may want to filter Pods further with other labelSelector with that key.
+
+	// labelKeysMap is keyed by label key and valued by the index of label key in labelKeys.
+	if labelSelector != nil {
+		labelKeysMap := map[string]int{}
+		for i, key := range matchLabelKeys {
+			labelKeysMap[key] = i
+		}
+		labelSelectorKeys := sets.New[string]()
+		for key := range labelSelector.MatchLabels {
+			labelSelectorKeys.Insert(key)
+		}
+		for _, matchExpression := range labelSelector.MatchExpressions {
+			key := matchExpression.Key
+			if i, ok := labelKeysMap[key]; ok && labelSelectorKeys.Has(key) {
```

Comment:
My memory of labelSelector validation is that the `.MatchLabels` and `.MatchExpressions` are managed independently.  so you can have a labelSelector with MatchLabels and MatchExpressions that check different labels.  See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go#L56-L66

Does something force sameness inside of affinity terms?